### ISSUE TITLE
BasicAnnotationProcessor does not support 'circular references' in annotations

### ISSUE THREAD

**Issue by facboy** [google/auto#780]
I have an annotation processor that generates a Dagger module off method annotations, and it's kind of neat to have the enclosing class itself be a `@Module` that includes the generated Dagger module, if that makes sense

```
@Module(includes = EnclosingModuleGeneratedModule.class)
abstract static class EnclosingModule {
    @DoGenerate
    abstract Thing thingTemplate();
}
```

At the moment this won't work with `BasicAnnotationProcessor` because it will never pass the `Elements` to its `ProcessingStep` because it cannot resolve the referenced class that the annotation processor is going to generate.

Is this something that it makes sense to address?  As I said, I find it kind of neat, but equally could be convinced that it's a bad idea generally.

FWIW, `AutoValue` seems to encourage this 'self-referencing' pattern, eg in the examples at https://github.com/google/auto/blob/master/value/userguide/index.md

**Comment by eamonnmcmanus** [comment]
Could you say a bit more? Which specific AutoValue examples are you referring to?

For what it's worth, AutoValue doesn't use BasicAnnotationProcessor. The [missing-types check](https://github.com/google/auto/blob/78eef202297404a65d4e4745315aa1c0b0efb3a7/common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java#L74) is too broad and would lead to circular dependencies being detected that don't actually matter. AutoValue does have [logic](https://github.com/google/auto/blob/6dfa04e3b636a5cf0ea3ec263b4fc991244f6fcb/value/src/main/java/com/google/auto/value/processor/AutoValueOrOneOfProcessor.java#L315) to defer processing in the face of missing types, but only when those types are needed for AutoValue to do its job.

**Comment by facboy** [comment]
oh, well this for example:

```
import com.google.auto.value.AutoValue;

@AutoValue
abstract class Animal {
  static Animal create(String name, int numberOfLegs) {
    return new AutoValue_Animal(name, numberOfLegs);
  }

  abstract String name();
  abstract int numberOfLegs();
}
```

I believe this wouldn't bother `BasicAnnotationProcessor` because the missing type is inside a method?

**Comment by eamonnmcmanus** [comment]
AutoValue is actually unable to care about missing types referenced from within method code, because the annotation processing API doesn't include the ability to see method code. So this isn't relevant to BasicAnnotationProcessor either.

I'm still finding the initial description a bit hard to follow. Is the cited code something that your processor is generating, something that another processor is generating, or something that the user writes? Perhaps you could spell out exactly what you are doing and how it fails.

**Comment by facboy** [comment]
The cited code in the OP is something that the user writes, my annotation processor processes it to generate `EnclosingModuleGeneratedModule`.  My annotation processor can't inherit from `BasicAnnotationProcessor` because the 'to-be-generated' `EnclosingModuleGeneratedModule` class cannot be resolved so `BasicAnnotationProcessor` will never allow the `ProcessingSteps` to process `EnclosingModule`.

**Comment by eamonnmcmanus** [comment]
OK, well the first thing to note is that your annotation processor is not going to be able to see what the name is in this line:

```
@Module(includes = EnclosingModuleGeneratedModule.class)
```

if it is supposed to generate that class. There's no standard way to see what the name of referenced class is until that class has been defined. But let's assume your processor always generates `FooGeneratedModule` for input class `Foo`, the same way AutoValue always generates `AutoValue_Foo`. Now I think you would need a way to tell `BasicAnnotationProcessor` to ignore specific missing types, in this case the value given to `@Module(includes)`. I can't see a way of doing that that wouldn't be either very specific to this problem or very complicated. Depending on what your processor does, it might be that it doesn't need to handle deferred processing at all. In other words, are there types referenced in the code being compiled that your processor might need in order to do its work, and that might themselves be generated by other processors? Even if so, you're probably better off handling deferred processing yourself. The [AutoValue logic](https://github.com/google/auto/blob/6dfa04e3b636a5cf0ea3ec263b4fc991244f6fcb/value/src/main/java/com/google/auto/value/processor/AutoValueOrOneOfProcessor.java#L315) I mentioned before is a fairly small amount of code.

I'm going to close this issue, but if after trying it out you have a specific suggestion for how `BasicAnnotationProcessor` might handle cases like these and others cleanly, feel free to reopen.

**Comment by facboy** [comment]
yeah that's fair, i was wondering if it sounded like the kind of thing `BasicAnnotationProcessor` would/could handle cleanly.  at the moment i'm just handling the deferred processing similarly to AutoValue.
